src/SarifCli/Rules/SarifValidationSkimmerBase.cs

```diff
@@ -0,0 +1,331 @@
+using System.Resources;
+using Microsoft.CodeAnalysis.Sarif.Driver;
+using Microsoft.CodeAnalysis.Sarif.Readers;
+using Microsoft.Json.Pointer;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.CodeAnalysis.Sarif.Cli.Rules
+{
+    public abstract class SarifValidationSkimmerBase : SkimmerBase<SarifValidationContext>
+    {
+        private const string SarifSpecUri =
+            "https://rawgit.com/sarif-standard/sarif-spec/master/Static%20Analysis%20Results%20Interchange%20Format%20(SARIF).html";
+
+        private readonly Uri _defaultHelpUri = new Uri(SarifSpecUri);
+        private JToken _rootToken;
+
+        public override Uri HelpUri => _defaultHelpUri;
+
+        protected SarifValidationContext Context { get; private set; }
+
+        protected override sealed ResourceManager ResourceManager => RuleResources.ResourceManager;
+
+        public override sealed void Analyze(SarifValidationContext context)
+        {
+            Context = context;
+
+            string logContents = File.ReadAllText(context.TargetUri.LocalPath);
+            _rootToken = JToken.Parse(logContents);
+
+            JsonSerializerSettings settings = new JsonSerializerSettings
+            {
+                ContractResolver = SarifContractResolver.Instance
+            };
+
+            SarifLog log = JsonConvert.DeserializeObject<SarifLog>(logContents, settings);
+            string logPointer = string.Empty;
+
+            Visit(log, logPointer);
+        }
+
+        private void Visit(SarifLog log, string logPointer)
+        {
+            if (log.Runs != null)
+            {
+                Run[] runs = log.Runs.ToArray();
+                string runsPointer = logPointer.AtProperty(SarifPropertyName.Runs);
+
+                for (int iRun = 0; iRun < runs.Length; ++iRun)
+                {
+                    Run run = runs[iRun];
+                    string runPointer = runsPointer.AtIndex(iRun);
+
+                    Visit(run, runPointer);
+                }
+            }
+        }
+
+        private void Visit(AnnotatedCodeLocation annotatedCodeLocation, string annotatedCodeLocationPointer)
+        {
+            if (annotatedCodeLocation.PhysicalLocation != null)
+            {
+                string physicalLocationPointer = annotatedCodeLocationPointer.AtProperty(SarifPropertyName.PhysicalLocation);
+                Visit(annotatedCodeLocation.PhysicalLocation, physicalLocationPointer);
+            }
+        }
+
+        private void Visit(CodeFlow codeFlow, string codeFlowPointer)
+        {
+            if (codeFlow.Locations != null)
+            {
+                AnnotatedCodeLocation[] annotatedCodeLocations = codeFlow.Locations.ToArray();
+                string annotatedCodeLocationsPointer = codeFlowPointer.AtProperty(SarifPropertyName.Locations);
+
+                for (int iAnnotatedCodeLocation = 0; iAnnotatedCodeLocation < annotatedCodeLocations.Length; ++iAnnotatedCodeLocation)
```

Comment:
Originally, I didn't have the visiting code so nicely factored, so I had multiple loops (say, a loop over results, with a nested loop over locations). So I used the verbose index names for (what I fondly believed to be) clarity. I'll shorten them.

---
In reply to: [77240006](https://github.com/Microsoft/sarif-sdk/pull/501#discussion_r77240006) [](ancestors = 77240006)